src/io/file_descriptor.cr

```diff
@@ -238,97 +182,29 @@ class IO::FileDescriptor
   end
 
   private def unbuffered_read(slice : Bytes)
-    count = slice.size
-    loop do
-      bytes_read = LibC.read(@fd, slice.pointer(count).as(Void*), count)
-      if bytes_read != -1
-        return bytes_read
-      end
-
-      if Errno.value == Errno::EAGAIN
-        wait_readable
-      else
-        raise Errno.new "Error reading file"
-      end
-    end
-  ensure
-    if (readers = @readers) && !readers.empty?
-      add_read_event
+    read_syscall_helper(slice, "Error reading file") do
+      # `to_i32` is acceptable because `Slice#size` is a Int32
+      LibC.read(@fd, slice, slice.size).to_i32
```

Comment:
Why do we need this explicit conversion?

Reply:
Because read_syscall_helper is explicitly annotated to return i32.

Reply:
Yeah, but `read_syscall_helper` didn't exist before this refactor, so why does it need to be explicitly annotated?

Reply:
Because `IO#read` should always return `typeof(Bytes.new.size)`. That wasn't the case before.

Reply:
Got it :). Thanks for bearing with me!

Reply:
It would be fantastic if we could force return types from an abstract declaration. Returning anything other than i32 from read doesn't make sense, and a return value from write is similarly useless.

Reply:
Yeah, we were discussing exactly that with @matiasgarciaisaia in relation to https://github.com/crystal-lang/crystal/issues/4864).